Java/parkservices/src/main/java/com/amazon/randomcutforest/extendedrandomcutforest/threshold/BasicThresholder.java

```diff
@@ -0,0 +1,266 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.randomcutforest.extendedrandomcutforest.threshold;
+
+
+import static com.amazon.randomcutforest.CommonUtils.checkArgument;
+
+import com.amazon.randomcutforest.extendedrandomcutforest.threshold.state.BasicThresholderState;
+
+public class BasicThresholder implements IThresholder{
+
+    public static int IS_ANOMALY = 1;
+
+    public static int NOT_ANOMALY = 0;
+
+    protected double elasticity = 0.01;
+
+    protected int count = 0;
+
+    protected double primaryDiscount;
+
+    protected double secondaryDiscount;
+
+    // horizon = 0 is short term, switches to secondary
+    // horizon = 1 long term, switches to primary
+    protected double horizon = 0.5;
+
+    protected int minimumScores = 10;
+
+    protected Deviation primaryDeviation;
+
+    protected Deviation secondaryDeviation;
+
+    // fraction of the grade that comes from absolute scores in the long run
+    protected double absoluteScoreFraction = 0.5;
+
+    // the upper threshold of scores above which points are likely anomalies
+    protected double upperThreshold = 2.0;
+    // the upper threshold of scores above which points are likely anomalies
+    protected double lowerThreshold = 1.0;
+    //initial absolute threshold used to determine anomalies before sufficient values are seen
+    protected double initialThreshold = 1.5;
+    // used to determine the suprise coefficient above which we can call a potential anomaly
+    protected double zFactor = 2.5;
+    // an upper bound of zFactor and triggerFactor beyond which the point is mathematically anomalous
+    // is useful in determining grade
+    protected double upperZfactor = 5.0;
+
+    protected double lastScore;
+
+
+    public BasicThresholder(double discount){
+        this.primaryDiscount=discount;
+        primaryDeviation = new Deviation(discount);
+        this.secondaryDiscount=discount;
+        secondaryDeviation = new Deviation(discount);
+    }
+
+    public BasicThresholder(Deviation deviation){
+        primaryDiscount = deviation.getDiscount();
+        secondaryDiscount = primaryDiscount;
+        this.primaryDeviation = deviation;
+        this.secondaryDeviation = new Deviation(deviation.getDiscount());
+    }
+
+    public BasicThresholder(BasicThresholderState state, Deviation primary, Deviation secondary){
```

Comment:
Shouldn't we be using a mapper class to construct a BasicThresholder from a BasicThresholderState and not a public constructor?

Reply:
We are using mappers -- it was getting too cumbersome to have so many parameters, and I passed the state. We have used constructors for pointstore before -- we can use that, or have a "parameters" class which is separate from state. I like the parameters approach over the constructors with 30 options ...

Reply:
Fixed.